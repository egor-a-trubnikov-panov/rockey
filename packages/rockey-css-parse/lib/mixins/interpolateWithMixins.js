import isFunction from 'lodash/isFunction';
import camelCase from 'lodash/camelCase';

const MIXIN_PREFIX = '|MIXIN_';
const getMixinName = mixinFunc => mixinFunc.displayName || mixinFunc.name;

const interpolateWithMixins = (strings, ...values) => {
  const mixinsFunctions = {};
  let counter = 0;

  const raw = strings.reduce((rule, part, i) => {
    let value = values[i] === undefined ? '' : values[i];
    let raw = part.trim();

    let append = value;

    if (isFunction(value)) {
      let found = false;
      let i = 0;
      for (i = raw.length - 1; i > 0; i--) {
        if (!found) {
          if (raw[i] === ':') {
            found = true;
          } else if (raw[i] !== ' ') {
            break;
          }
        }

        if (found) {
          if (raw[i] === ' ' || raw[i] === ';' || raw[i] === '{') {
            break;
          }
        }
      }

      let prop = null;
      let name = null;

      if (found) {
        const index = i ? i : 0;

        prop = raw.slice(index, -1).trim();
        raw = raw.slice(0, index);
        name = camelCase(prop);
      } else {
        name = getMixinName(value);
      }

      let placeholder = `${MIXIN_PREFIX}${name}${++counter}`;

      if (found) {
        mixinsFunctions[placeholder] = (...args) => {
          const propValue = value(...args);
          return propValue
            ? {
                [prop]: propValue,
              }
            : null;
        };

        mixinsFunctions[placeholder].prop = prop;
        mixinsFunctions[placeholder].displayName = name;
      } else {
        mixinsFunctions[placeholder] = value;
      }

      append = `${placeholder};`;
      // append = placeholder + '; ';
    }

    return rule + raw + ' ' + append;
  }, '');

  return { raw, mixinsFunctions };
};

export default interpolateWithMixins;
