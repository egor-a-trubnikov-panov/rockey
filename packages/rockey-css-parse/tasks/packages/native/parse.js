import { argv } from 'yargs';
import chalk from 'chalk';

import getNativeCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import createParser from '../../../lib/parse';
import { stringifyRules } from '../../../lib/stringify';

const size = argv.size || 1000;
const css = getNativeCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

const parse = createParser();

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
