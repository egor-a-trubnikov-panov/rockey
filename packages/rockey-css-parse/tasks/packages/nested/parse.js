import { argv } from 'yargs';
import chalk from 'chalk';

import getCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import createParser from '../../../lib/parse';
import { stringifyRules } from '../../../lib/stringify';

const size = argv.size || 10;
const css = getCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

const parse = createParser();

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
