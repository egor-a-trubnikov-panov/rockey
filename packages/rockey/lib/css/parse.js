// const validateRule = (rule, value) => {};
// const autoprefixRule = (rule, value) => {};

const cache = new Map();

const parseCss = styles => {
  const stylesParts = styles
    .trim()
    .split(';')
    .map(s => s ? s.split(':') : null)
    .filter(s => !!s);

  const key = stylesParts.toString();

  if (!key.trim()) {
    return {};
  }

  if (cache.has(key)) {
    return cache.get(key);
  }

  const stylesJson = stylesParts.reduce(
    (json, value) => {
      if (value) {
        const key = value[0].trim();
        if (json[key]) {
          // css fallback values
          let prev = Array.isArray(json[key]) ? json[key] : [json[key]];

          json[value[0].trim()] = prev.concat([
            value.slice(1).join(':').trim(),
          ]);
        } else {
          json[value[0].trim()] = value.slice(1).join(':').trim();
        }

        // validateRule(value[0], json[value[0]]);
        // json[value[0].trim()] = autoprefixRule(value.slice(1).join(':').trim());
      }
      return json;
    },
    {}
  );

  cache.set(key, stylesJson);

  return stylesJson;
};

// @replace-start
const isModificatorStartSymbol = symbol => {
  return symbol === '@' || symbol === ':' || symbol === '[';
};
// @replace-end

// @replace-start
const isModificatorEndSymbol = symbol => {
  return symbol === '{' || symbol === ';';
};
// @replace-end

// TODO: make this func beauty :)
const isStartsWithModificator = raw => {
  return 0 === raw.indexOf('@media') ||
    0 === raw.indexOf('@keyframes') ||
    0 === raw.indexOf('::after') ||
    0 === raw.indexOf('::before') ||
    0 === raw.indexOf('::first-letter') ||
    0 === raw.indexOf('::first-line') ||
    0 === raw.indexOf(':active') ||
    0 === raw.indexOf(':checked') ||
    0 === raw.indexOf(':disabled') ||
    0 === raw.indexOf(':empty') ||
    0 === raw.indexOf(':enabled') ||
    0 === raw.indexOf(':first-child') ||
    0 === raw.indexOf(':first-of-type') ||
    0 === raw.indexOf(':focus') ||
    0 === raw.indexOf(':hover') ||
    0 === raw.indexOf(':in-range') ||
    0 === raw.indexOf(':invalid') ||
    0 === raw.indexOf(':lang') ||
    0 === raw.indexOf(':last-child') ||
    0 === raw.indexOf(':last-of-type') ||
    0 === raw.indexOf(':link') ||
    0 === raw.indexOf(':not') ||
    0 === raw.indexOf(':nth-child') ||
    0 === raw.indexOf(':nth-last-child') ||
    0 === raw.indexOf(':nth-last-of-type') ||
    0 === raw.indexOf(':nth-of-type') ||
    0 === raw.indexOf(':only-of-type') ||
    0 === raw.indexOf(':only-child') ||
    0 === raw.indexOf(':optional') ||
    0 === raw.indexOf(':out-of-range') ||
    0 === raw.indexOf(':read-only') ||
    0 === raw.indexOf(':read-write') ||
    0 === raw.indexOf(':required') ||
    0 === raw.indexOf(':root') ||
    0 === raw.indexOf(':target') ||
    0 === raw.indexOf(':valid') ||
    0 === raw.indexOf(':visited');
};

const parse = (raw, parent) => {
  let openedBrackets = 0;
  let openedModificatorBrackets = 0;
  let styles = '';

  let component = null;
  let combinedComponents = [];
  let components = {};

  let possibleModificator = 0;
  let modificators = {};
  let modificator = '';

  let current = '';
  let currentBackup = '';

  // @replace-start
  const restore = () => {
    current = currentBackup + current;
    currentBackup = '';
  };
  // @replace-end

  // @replace-start
  const clearAndRestore = () => {
    current = currentBackup;
    currentBackup = '';
  };
  // @replace-end

  // @replace-start
  const backup = () => {
    currentBackup = current;
    current = '';
  };
  // @replace-end

  // @replace-start
  const possibleModificatorStart = () => {
    possibleModificator = true;
  };
  // @replace-end

  // @replace-start
  const isPossibleModificator = () => {
    return possibleModificator;
  };
  // @replace-end

  // @replace-start
  const possibleModificatorEnd = () => {
    possibleModificator = false;
  };
  // @replace-end

  // @replace-start
  const isModificator = () => {
    return isStartsWithModificator(current.trim());
  };
  // @replace-end

  // @replace-start
  const openBracket = () => {
    openedBrackets++;
  };
  // @replace-end

  // @replace-start
  const closeBracket = () => {
    openedBrackets--;
  };
  // @replace-end

  // @replace-start
  const isAllBracketsClosed = () => {
    return openedBrackets === 0;
  };
  // @replace-end

  // @replace-start
  const openModificatorBracket = () => {
    openedModificatorBrackets++;
  };
  // @replace-end

  // @replace-start
  const closeModificatorBracket = () => {
    openedModificatorBrackets--;
  };
  // @replace-end

  // @replace-start
  const isAllModificatorBracketsClosed = () => {
    return openedModificatorBrackets === 0;
  };
  // @replace-end

  // @replace-start
  const startModificator = () => {
    modificator = current;
    current = '';
  };
  // @replace-end

  // @replace-start
  const saveModificator = () => {
    modificator = modificator.trim();

    // @remove-start
    if (modificators[modificator]) {
      throw new Error(
        `(parse error) "${modificator}" duplicated definiton at one block`
      );
    }
    // @remove-end

    modificators[modificator] = parse(current, {
      parentType: 'modificator',
      name: modificator,
    });

    modificator = '';
  };
  // @replace-end

  // @replace-start
  const isInsideModificator = () => {
    return !!modificator;
  };
  // @replace-end

  // @replace-start
  const isRootComponentBlockStarts = symbol => {
    return symbol === '{' && isAllBracketsClosed();
  };
  // @replace-end

  // @replace-start
  const startComponent = () => {
    let parts = current.trim().split(' ');
    if (parts.length === 1) {
      component = parts[0];
    } else {
      combinedComponents = [];

      let i = 0;
      for (i = parts.length - 1; i >= 0; i--) {
        const part = parts[i];
        const charCode = part.charCodeAt(0);

        /* All chars starts with uppercase or xxx% for animations  */
        if (
          (charCode >= 65 && charCode <= 90) || part[part.length - 1] === '%'
        ) {
          combinedComponents.push(part.replace(',', '').trim());
        } else if (part[0] === '~' || part[0] === '+') {
          let size = combinedComponents.length;
          combinedComponents[size - 1] = part[0] + combinedComponents[size - 1];
        } else {
          break;
        }
      }

      if (combinedComponents.length) {
        component = combinedComponents[0];
        combinedComponents = combinedComponents.slice(1);
        parts = parts.slice(0, i + 1);
      } else {
        component = parts[parts.length - 1];
        parts = parts.slice(0, -1);
      }

      styles += parts.join(' ');
    }

    current = '';
  };
  // @replace-end

  // @replace-start
  const isInsideComponent = () => {
    return !!component;
  };
  // @replace-end

  // @replace-start
  const save = symbol => {
    current += symbol;
  };
  // @replace-end

  // @replace-start
  const saveComponent = () => {
    // @remove-start
    if (components[component]) {
      throw new Error(
        `(parse error) "${component}" duplicated definiton at one block`
      );
    }
    // @remove-end
    components[component] = parse(current, {
      parentType: 'component',
      name: component,
      combinedComponents,
    });

    combinedComponents = [];

    component = null;
    current = '';
  };
  // @replace-end

  // @replace-start
  const updateStylesTail = () => {
    if (current) {
      styles += current;
    }
  };
  // @replace-end

  // ---
  const mixins = [];
  let possibleMixin = false;
  let mixin = '';

  // @replace-start
  const isMixinStartSymbol = symbol => {
    return symbol === '_';
  };
  // @replace-end

  // @replace-start
  const isMixinEndSymbol = symbol => {
    return symbol === ' ';
  };
  // @replace-end

  // @replace-start
  const possibleMixinStart = () => {
    possibleMixin = true;
  };
  // @replace-end

  // @replace-start
  const possibleMixinEnd = () => {
    possibleMixin = false;
  };
  // @replace-end

  // @replace-start
  const isPossibleMixin = () => {
    return possibleMixin;
  };
  // @replace-end

  // @replace-start
  const saveMixinString = symbol => {
    mixin += symbol;
  };
  // @replace-end

  // @replace-start
  const clearMixin = () => {
    mixin = '';
  };
  // @replace-end

  // @replace-start
  const removeMixinFromStyles = () => {
    current = current.replace(mixin.trim(), '');
  };
  // @replace-end

  // @replace-start
  const saveMixin = () => {
    mixins.push(mixin.trim());
  };
  // @replace-end

  // @remove-start
  const isInValidMixinPosition = () => {
    return !parent ||
      (currentBackup.trim().length &&
        !new RegExp(`(;|{)\\s+?${mixin.trim()}`).test(`${current}`.trim()));
  };
  // @remove-end

  // ----

  for (let symbol of raw) {
    if (!isInsideComponent() && !isInsideModificator()) {
      if (isMixinStartSymbol(symbol)) {
        possibleMixinStart();
      }

      if (isPossibleMixin()) {
        saveMixinString(symbol);
        if (isMixinEndSymbol(symbol)) {
          possibleMixinEnd();
          // @remove-start
          if (isInValidMixinPosition()) {
            if (parent) {
              const nearestCode = `${currentBackup}${current}`
                .replace(/\s+/g, ' ')
                // .replace(mixin.trim(), '{{ MIXIN }}' + mixin)
                .trim();

              console.warn(
                'seems wrong mixin position',
                `${parent.name} - ${nearestCode}`
              );
            } else {
              console.warn('mixin should not be at root');
            }
          }
          // @remove-end

          saveMixin();
          removeMixinFromStyles();
          clearMixin();
        }
      }
    }

    if (!isInsideComponent() && !isInsideModificator()) {
      if (!isPossibleModificator() && isModificatorStartSymbol(symbol)) {
        possibleModificatorStart();
        backup();
      }

      if (isPossibleModificator() && isModificatorEndSymbol(symbol)) {
        if (isModificator()) {
          startModificator();
        } else {
          restore();
        }

        possibleModificatorEnd();
      }
    }

    if (
      !isPossibleModificator() &&
      !isInsideModificator() &&
      isRootComponentBlockStarts(symbol)
    ) {
      openBracket();
      startComponent();
      // continue;
    } else if (symbol === '{') {
      if (isInsideComponent()) {
        openBracket();

        if (openedBrackets !== 1) {
          save(symbol);
        }
      } else if (isInsideModificator()) {
        openModificatorBracket();

        if (openedModificatorBrackets !== 1) {
          save(symbol);
        }
      }
    } else if (symbol === '}') {
      if (isInsideComponent()) {
        closeBracket();
      } else if (isInsideModificator()) {
        closeModificatorBracket();
      }

      if (isInsideComponent()) {
        if (isAllBracketsClosed()) {
          saveComponent();
        } else {
          save(symbol);
        }
      } else if (isInsideModificator()) {
        if (isAllModificatorBracketsClosed()) {
          saveModificator();
          clearAndRestore();
        } else {
          save(symbol);
        }
      }
    } else {
      save(symbol);
    }
  }

  restore();
  updateStylesTail();

  return {
    mixins,
    components,
    combinedComponents: (parent && parent.combinedComponents) || [],
    styles: parseCss(styles),
    modificators,
  };
};

export default inline => {
  return parse(inline.replace(/\r|\n/g, '').replace(/\s+/g, ' '));
};
