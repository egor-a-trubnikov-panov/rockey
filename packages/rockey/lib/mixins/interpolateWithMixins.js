import isFunction from 'lodash/isFunction';

let mixinCounter = 0;

const MIXIN_PREFIX = '_MIXIN_';

const getMixinName = mixin =>
  `${mixin.displayName || mixin.name || `${++mixinCounter}`}`;

const interpolateWithMixins = (strings, ...values) => {
  const mixinsFunctions = {};
  let counter = 0;

  const raw = strings.reduce(
    (rule, part, i) => {
      let value = values[i] === undefined ? '' : values[i];

      if (isFunction(value)) {
        const name = getMixinName(value);

        let placeholder = `${MIXIN_PREFIX}${name}${++counter}`;
        mixinsFunctions[placeholder] = value;

        value = placeholder;
      }

      return rule + part + value;
    },
    ''
  );

  return { raw, mixinsFunctions };
};

export default interpolateWithMixins;

export const addMixins = (raw, mixinsFunctions, mixins) => {
  mixins.forEach(mixin => {
    const name = `${MIXIN_PREFIX}_ADDED_${getMixinName(mixin)}`;
    raw = raw + ' ' + name;
    mixinsFunctions[name] = mixin;
  });

  return raw;
};
