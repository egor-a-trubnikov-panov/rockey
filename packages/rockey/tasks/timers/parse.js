import { argv } from 'yargs';
import chalk from 'chalk';

import getCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import parse from '../../lib/css/parse';
import { stringifyRules } from '../../lib/styleSheets/utils/stringify';

const size = argv.size || 10;
const css = getCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
