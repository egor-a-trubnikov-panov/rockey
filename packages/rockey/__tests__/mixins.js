import parse from '../lib/css/parse';
import interpolateWithMixins from '../lib/mixins/interpolateWithMixins';

console.warn = jest.fn();

const primary = props => {
  if (props.isPrimary) {
    return `color: blue`;
  }
};

const isRaised = props => {
  if (props.isPrimary) {
    return `color: black`;
  }
};

describe('mixins', () => {
  beforeEach(() => {
    console.warn.mockClear();
  });

  it('1 mixin', () => {
    const { raw } = interpolateWithMixins`
      Button {
        color: yellow;

        ${primary}
      }
    `;

    const generated = parse(raw);
    expect(generated.components.Button.mixins.length).toEqual(1);
  });

  // it('2 mixins', () => {
  //   const { raw } = interpolateWithMixins`
  //     Button {
  //       color: yellow;
  //
  //       ${primary}
  //       ${isRaised}
  //     }
  //   `;
  //
  //    parse(raw);
  //
  //   const classNameMap = generated.classNameMap;
  //   expect(generated.components.Button.mixins.length).toEqual(2);
  // });

  it('wrong mixin position #1 (named mixin)', () => {
    const { raw } = interpolateWithMixins`
      Button {
        color: red;
        background: ${isRaised}
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(1);
    expect(console.warn.mock.calls[0][1]).toEqual(
      'Button - color: red; background: _MIXIN_isRaised1'
    );
  });

  it('wrong mixin position #2', () => {
    const { raw } = interpolateWithMixins`
      Button {
        color: red;

        Icon {
          background: ${primary}
        }
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(1);
    expect(console.warn.mock.calls[0][1]).toEqual(
      'Icon - background: _MIXIN_primary1'
    );
  });

  it('wrong mixin position #3', () => {
    const { raw } = interpolateWithMixins`
      Button {
        color: red;

        :hover {
          background: ${primary}
        }
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(1);
    expect(console.warn.mock.calls[0][1]).toEqual(
      ':hover - background: _MIXIN_primary1'
    );
  });

  it('wrong mixin position #4', () => {
    const { raw } = interpolateWithMixins`
      ${primary}

      Button {
        color: red;
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(1);
    expect(console.warn.mock.calls[0][1]).toEqual(
      'mixin should not be at root'
    );
  });

  it('wrong mixin position #5', () => {
    const { raw } = interpolateWithMixins`

      Button {
        color: red;
      }

      ${primary}
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(1);
    expect(console.warn.mock.calls[0][1]).toEqual(
      'mixin should not be at root'
    );
  });

  it('correct mixin position #1', () => {
    const { raw } = interpolateWithMixins`
      Button {
        color: red;

        :hover {
          ${primary}
        }
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(0);
  });

  it('correct mixin position #2', () => {
    const { raw } = interpolateWithMixins`
      Button {
        color: red;

        ${primary}
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(0);
  });

  it('correct mixin position #3', () => {
    const { raw } = interpolateWithMixins`
      Button {
        ${primary}

        color: red;
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(0);
  });

  it('correct mixin position #4', () => {
    const { raw } = interpolateWithMixins`
      Button {

        @media(min-width: 300px) {
          ${primary}
        }

        color: red;
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(0);
  });

  it('correct mixin position #5', () => {
    const { raw } = interpolateWithMixins`
      Button {

        @media(min-width: 300px) {
          color: red;
          ${primary}
        }
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(0);
  });

  it('correct mixin position #6', () => {
    const { raw } = interpolateWithMixins`
      Button {

        @media(min-width: 300px) {
          color: red;
          :hover {
            Icon {
              ${primary}
            }
          }
        }
      }
    `;

    parse(raw);

    expect(console.warn.mock.calls.length).toEqual(0);
  });

  // ---
  // it('mixins results should contain only strings', () => {
  //   const { raw } = interpolateWithMixins`
  //     Button {
  //       color: red
  //       ${() => {
  //         return `
  //           color: blue;
  //           ${() => {}}
  //         `
  //       }}
  //     }
  //   `;
  //
  //    parse(raw);
  //
  //   expect(console.warn.mock.calls.length).toEqual(0);
  // });
});
