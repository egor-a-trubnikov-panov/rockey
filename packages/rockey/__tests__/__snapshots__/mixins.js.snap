// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule 1`] = `
Object {
  "Button": "Button-hash m-when-hash-1 m-background-hash-1 m-CustomName-hash-1",
}
`;

exports[`rule 2`] = `
Array [
  Array [
    Array [
      Object {
        "media": null,
        "mixins": Array [
          [Function],
          [Function],
          [Function],
        ],
        "root": Array [
          "Button",
        ],
        "selector": Array [
          ".Button-hash .A-hash .B-hash:active .C-hash .D-hash .E-hash:hover",
          ".Button-hash .A-hash .B1-hash .C-hash .D-hash .E-hash:hover",
          ".Button-hash .A-hash .B2-hash .C-hash .D-hash .E-hash:hover",
          ".Button-hash .A-hash .B3-hash .C-hash .D-hash .E-hash:hover",
        ],
        "styles": Object {
          "color": "green",
        },
      },
      Object {
        "media": null,
        "mixins": Array [],
        "root": Array [
          "Button",
        ],
        "selector": Array [
          ".Button-hash",
        ],
        "styles": Object {
          "color": "red",
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "selector": Array [
          ".m-when-hash-1.Button-hash .A-hash .B-hash:active .C-hash .D-hash .E-hash:hover",
          ".m-when-hash-1.Button-hash .A-hash .B1-hash .C-hash .D-hash .E-hash:hover",
          ".m-when-hash-1.Button-hash .A-hash .B2-hash .C-hash .D-hash .E-hash:hover",
          ".m-when-hash-1.Button-hash .A-hash .B3-hash .C-hash .D-hash .E-hash:hover",
        ],
        "styles": Object {
          "color": "red",
        },
      },
      Object {
        "selector": Array [
          ".m-background-hash-1.Button-hash .A-hash .B-hash:active .C-hash .D-hash .E-hash:hover",
          ".m-background-hash-1.Button-hash .A-hash .B1-hash .C-hash .D-hash .E-hash:hover",
          ".m-background-hash-1.Button-hash .A-hash .B2-hash .C-hash .D-hash .E-hash:hover",
          ".m-background-hash-1.Button-hash .A-hash .B3-hash .C-hash .D-hash .E-hash:hover",
        ],
        "styles": Object {
          "background": "yellow",
        },
      },
      Object {
        "selector": Array [
          ".m-CustomName-hash-1.Button-hash .A-hash .B-hash:active .C-hash .D-hash .E-hash:hover",
          ".m-CustomName-hash-1.Button-hash .A-hash .B1-hash .C-hash .D-hash .E-hash:hover",
          ".m-CustomName-hash-1.Button-hash .A-hash .B2-hash .C-hash .D-hash .E-hash:hover",
          ".m-CustomName-hash-1.Button-hash .A-hash .B3-hash .C-hash .D-hash .E-hash:hover",
        ],
        "styles": Object {
          "font-size": "20px",
        },
      },
    ],
  ],
]
`;

exports[`rule 3`] = `
Object {
  "Button": "Button-hash m-background-hash-2",
}
`;

exports[`rule 4`] = `
Array [
  Array [
    Array [
      Object {
        "media": null,
        "mixins": Array [
          [Function],
          [Function],
          [Function],
        ],
        "root": Array [
          "Button",
        ],
        "selector": Array [
          ".Button-hash .A-hash .B-hash:active .C-hash .D-hash .E-hash:hover",
          ".Button-hash .A-hash .B1-hash .C-hash .D-hash .E-hash:hover",
          ".Button-hash .A-hash .B2-hash .C-hash .D-hash .E-hash:hover",
          ".Button-hash .A-hash .B3-hash .C-hash .D-hash .E-hash:hover",
        ],
        "styles": Object {
          "color": "green",
        },
      },
      Object {
        "media": null,
        "mixins": Array [],
        "root": Array [
          "Button",
        ],
        "selector": Array [
          ".Button-hash",
        ],
        "styles": Object {
          "color": "red",
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "selector": Array [
          ".m-when-hash-1.Button-hash .A-hash .B-hash:active .C-hash .D-hash .E-hash:hover",
          ".m-when-hash-1.Button-hash .A-hash .B1-hash .C-hash .D-hash .E-hash:hover",
          ".m-when-hash-1.Button-hash .A-hash .B2-hash .C-hash .D-hash .E-hash:hover",
          ".m-when-hash-1.Button-hash .A-hash .B3-hash .C-hash .D-hash .E-hash:hover",
        ],
        "styles": Object {
          "color": "red",
        },
      },
      Object {
        "selector": Array [
          ".m-background-hash-1.Button-hash .A-hash .B-hash:active .C-hash .D-hash .E-hash:hover",
          ".m-background-hash-1.Button-hash .A-hash .B1-hash .C-hash .D-hash .E-hash:hover",
          ".m-background-hash-1.Button-hash .A-hash .B2-hash .C-hash .D-hash .E-hash:hover",
          ".m-background-hash-1.Button-hash .A-hash .B3-hash .C-hash .D-hash .E-hash:hover",
        ],
        "styles": Object {
          "background": "yellow",
        },
      },
      Object {
        "selector": Array [
          ".m-CustomName-hash-1.Button-hash .A-hash .B-hash:active .C-hash .D-hash .E-hash:hover",
          ".m-CustomName-hash-1.Button-hash .A-hash .B1-hash .C-hash .D-hash .E-hash:hover",
          ".m-CustomName-hash-1.Button-hash .A-hash .B2-hash .C-hash .D-hash .E-hash:hover",
          ".m-CustomName-hash-1.Button-hash .A-hash .B3-hash .C-hash .D-hash .E-hash:hover",
        ],
        "styles": Object {
          "font-size": "20px",
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "selector": Array [
          ".m-background-hash-2.Button-hash .A-hash .B-hash:active .C-hash .D-hash .E-hash:hover",
          ".m-background-hash-2.Button-hash .A-hash .B1-hash .C-hash .D-hash .E-hash:hover",
          ".m-background-hash-2.Button-hash .A-hash .B2-hash .C-hash .D-hash .E-hash:hover",
          ".m-background-hash-2.Button-hash .A-hash .B3-hash .C-hash .D-hash .E-hash:hover",
        ],
        "styles": Object {
          "background": "green",
        },
      },
    ],
  ],
]
`;
